### PR TITLE
[JW8-11120] [JW8-11119] Add tizen-controls and controls loading

### PR DESCRIPTION
**Pull request by mamaddox**
### This PR will...
This is the basis for tizen-controls in the player. It removes/updates the code used in the first demo app including the following:
* Extracts tizen controls from controls.js into `tizen-controls`
* Adds `tizen-controlbar`
* Removes previously added remote handlers and `isTizenApp` checks that are no longer needed from:
   * menu.js
   * settings-menu.js
   * rightclick.js
   * view.js
* Removes previously added `:not(jw-tizen-app)`css selectors that are no longer needed from:
   * controlbar.less
   * settings-menu.less 
* Updates `core-bundle-loader` and `controls-loader` to load tizen controls when tizen environment is detected
* Removes isTizenApp model attribute in favor of using a tizen environment check
* Adds `tizenApp` to `browser` and `environment` to differentiate between tizen web apps and the tizen browser
*Note: `OS.tizen` is currently used as opposed to `OS.tizenApp` for developing because the tizen simulator uses the tizen browser user agent.

### Why is this Pull Request needed?
Tizen Demo App

### Are there any Pull Requests open in other repos which need to be merged with this?
https://github.com/jwplayer/jwplayer-commercial/pull/7657

#### Addresses Issue(s):

[JW8-11120](https://jwplayer.atlassian.net/browse/JW8-11120)
[JW8-11119](https://jwplayer.atlassian.net/browse/JW8-11119)

### Checklist
- [ ] Jenkins builds and unit tests are passing
- [ ] I have reviewed the automated results

### CODE REVIEW

File: src/css/controls/flags/tizen.less [hunk 1]
```diff
@@ -1,93 +1,239 @@
...
+            border-radius: 0;
+            background-size: auto;
+            box-shadow: none;
+            left: 0;
+            bottom: 0;
+            padding: inherit;
+        }
 
         .jw-text {
-            font-size: @tizen-font-size;
```
Comment (jnatalzia) [inline src/css/controls/flags/tizen.less:81]: are these vars still useful at all?

File: src/js/api/core-bundle-loader.js [hunk 2]
```diff
@@ -20,6 +21,10 @@ function selectBundle(model) {
     const polyfills = requiresPolyfills();
     const html5Provider = requiresProvider(model, 'html5');
 
+    if (OS.tizen) {
+        return loadWebCore();
```
Comment (jnatalzia) [inline src/js/api/core-bundle-loader.js:25]: nice

File: src/js/controller/controls-loader.js [hunk 3]
```diff
@@ -1,19 +1,31 @@
...
-        }, 'jwplayer.controls');
+        if (OS.tizen) {
+            controlsPromise = require.ensure(['view/controls/tizen/tizen-controls'], function (require) {
+                const ControlsModule = require('view/controls/tizen/tizen-controls').default;
+                ControlsLoader.controls = ControlsModule;
+                return ControlsModule;
+            }, function() {
+                controlsPromise = null;
+                chunkLoadWarningHandler(301130)();
+            }, 'jwplayer.controls.tizen');
```
Comment (jnatalzia) [inline src/js/controller/controls-loader.js:18]: this is awesome!

File: src/js/controller/model.ts [hunk 4]
```diff
@@ -285,7 +285,7 @@ class Model extends SimpleModel {
     // Desktop players must have autostart set to viewable
     setAutoStart(autoStart?: AutoStart): void {
         // Always autostart for tizen app
-        if (this.attributes.isTizenApp) {
+        if (OS.tizen) {
```
Comment (jnatalzia) [inline src/js/controller/model.ts:288]: 👍

File: src/js/view/controls/controls.js [hunk 5]
```diff
@@ -467,148 +457,6 @@ export default class Controls extends Events {
         model.set('controlsEnabled', true);
     }
 
-    /* Want to move to controls/tv/tizen-controls.js which extends controls.js */
-    enableTizenControls(api, model) {
```
Comment (jnatalzia) [inline src/js/view/controls/controls.js:471]: great refactor

File: src/js/view/controls/tizen/tizen-controlbar.ts [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+        appendChildren(elements.topContainer, topLayout);
+        appendChildren(elements.bottomContainer, bottomLayout);
+        appendChildren(this.el, layout);
+
+        // Initial State
+        elements.play.show();
+        elements.back.show();
+    }
+
+    getRightButton(activeButton: Button, containerLayout: ControlbarElement[]): Button | undefined {
```
Comment (jnatalzia) [inline src/js/view/controls/tizen/tizen-controlbar.ts:118]: i like this approach to dynamic button finding, i'm curious your thoughts on a way to potentially create an abstraction of `getRight` and `getLeft`. all of the logic is exactly the same except the for loop condition. it might be more trouble than it's worth but something to explore!

File: src/js/view/controls/tizen/tizen-controlbar.ts [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                break;
+            default:
+                break;
+        }
+    }
+
+    setActiveButton(nextButton: Button): void {
+        const currentActiveButton = this.activeButton;
+        if (currentActiveButton === nextButton) {
+            return;
```
Comment (jnatalzia) [inline src/js/view/controls/tizen/tizen-controlbar.ts:220]: nice

File: src/js/view/controls/tizen/tizen-controls.ts [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+import TizenControlbar from 'view/controls/tizen/tizen-controlbar';
+import DisplayContainer from 'view/controls/display-container';
+import NextUpToolTip from 'view/controls/nextuptooltip';
+import SettingsMenu from 'view/controls/components/menu/settings-menu.js';
+import { addClass } from 'utils/dom';
+import { STATE_PLAYING, STATE_PAUSED } from 'events/events';
+import type ViewModel from 'view/view-model';
+import type { PlayerAPI } from 'types/generic.type';
+
+const Controls = require('../controls').default;
```
Comment (jnatalzia) [inline src/js/view/controls/tizen/tizen-controls.ts:11]: what's the reasoning behind this syntax?

File: src/js/view/controls/tizen/tizen-controls.ts [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+    playerContainer: HTMLElement;
+    div: HTMLElement | null;
+    backdrop: HTMLElement | null;
+    pauseDisplay: HTMLElement | null;
+    displayContainer: DisplayContainer | null;
+    controlbar: TizenControlbar | null;
+    settingsMenu: SettingsMenu | null;
+    showing: boolean;
+    keydownCallback: ((evt: KeyboardEvent) => void) | null;
+    addControls: any;
```
Comment (jnatalzia) [inline src/js/view/controls/tizen/tizen-controls.ts:30]: i assume these would get re-typed when controls is typed? or maybe removed?

File: src/js/view/controls/tizen/tizen-controls.ts [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+        const element = this.context.createElement('div');
+        element.className = 'jw-tizen-controls jw-tizen-reset';
+        this.div = element;
+    
+        const backdrop = this.context.createElement('div');
+        backdrop.className = 'jw-controls-backdrop jw-reset';
+        this.backdrop = backdrop;
+
+        // Pause Display
+        if (!this.pauseDisplay) {
```
Comment (jnatalzia) [inline src/js/view/controls/tizen/tizen-controls.ts:61]: what are your thoughts on using templating for this similar to how we construct other complex elements? (https://github.com/jwplayer/jwplayer-commercial/blob/master/src/js/templates/dismissButton.js)

File: src/js/view/controls/tizen/tizen-controls.ts [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+        const localization = model.get('localization');
+        this.settingsMenu = new SettingsMenu(api, model.player, this.controlbar, localization);
+
+        const handleKeydown = (evt: KeyboardEvent) => {
+            if (this.controlbar) {
+                this.controlbar.handleKeydown(evt, this.showing);
+            }
+            switch (evt.keyCode) {
+                case 37: // left-arrow
+                    this.userActive();
```
Comment (jnatalzia) [inline src/js/view/controls/tizen/tizen-controls.ts:121]: you can remove all these cases and just make it a fall-through (unless out lint rules don't allow that)
```
case 37:
case 39:
...etc:
  this.userActive();
  break;
```

File: src/js/view/controls/tizen/tizen-controls.ts [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+                    this.removePlayer(api, model);
+                    break;
+                default:
+                    break;
+            }
+        };
+
+        // Trigger backClick when all videos complete      
+        api.on('playlistComplete', () => {
+            this.onBackClick(api, model);
```
Comment (jnatalzia) [inline src/js/view/controls/tizen/tizen-controls.ts:173]: nice

File: src/js/view/controls/tizen/tizen-controls.ts [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+        api.setFullscreen(true);
+
+        model.set('controlsEnabled', true);
+    }
+
+    disable(model: ViewModel): void {
+        if (this.keydownCallback) {
+            document.removeEventListener('keydown', this.keydownCallback);
+        }
+        super.disable.apply(this, [model]);
```
Comment (jnatalzia) [inline src/js/view/controls/tizen/tizen-controls.ts:191]: `super.disable.call(this, model)` stops an extra array from being created 👍

Comment (jnatalzia) [reply src/js/view/controls/tizen/tizen-controls.ts:191]: same comment on the below

File: src/js/view/controls/tizen/tizen-controlbar.ts (hunk 6 above)
Comment (mamaddox) [reply src/js/view/controls/tizen/tizen-controlbar.ts:118]: +1 I 100% would love to do this. I spent a little time trying to figure out a solution, but will definitely explore some more.

File: src/css/controls/flags/tizen.less (hunk 1 above)
Comment (mamaddox) [reply src/css/controls/flags/tizen.less:81]: good catch. not anymore.

File: src/js/view/controls/tizen/tizen-controls.ts (hunk 9 above)
Comment (mamaddox) [reply src/js/view/controls/tizen/tizen-controls.ts:30]: These should be removed when controls is typed since they are methods on controls/events. They were throwing a type error when called.

File: src/js/view/controls/tizen/tizen-controls.ts (hunk 8 above)
Comment (mamaddox) [reply src/js/view/controls/tizen/tizen-controls.ts:12]: Controls uses `require()` because there was an issue when using `import`. Because we resolve the import file paths by checking commercial first and then os, Controls was being imported as CommercialControls, which was then overwriting the TizenControls. Looking into some workarounds, though this may be resolved when controls get typed.

File: src/js/utils/browser.ts [hunk 14]
```diff
@@ -36,6 +36,10 @@ export function isTizen(): boolean {
     return userAgentMatch(/SMART-TV/);
 }
 
+export function isTizenApp(): boolean {
+    return isTizen() && userAgentMatch(/Version\//);
```
Comment (robwalch) [inline src/js/utils/browser.ts:40]: Could this be made a little more specific?

File: src/js/controller/controls-loader.js [hunk 15]
```diff
@@ -1,19 +1,31 @@
...
-            chunkLoadWarningHandler(301130)();
-        }, 'jwplayer.controls');
+        if (OS.tizen) {
+            controlsPromise = require.ensure(['view/controls/tizen/tizen-controls'], function (require) {
+                const ControlsModule = require('view/controls/tizen/tizen-controls').default;
+                ControlsLoader.controls = ControlsModule;
+                return ControlsModule;
+            }, function() {
+                controlsPromise = null;
+                chunkLoadWarningHandler(301130)();
```
Comment (robwalch) [inline src/js/controller/controls-loader.js:17]: There should be a different error code per file (so we know it's jwplayer.controls.tizen.js that failed to load).

File: src/js/view/controls/tizen/tizen-controls.ts (hunk 8 above)
Comment (robwalch) [reply src/js/view/controls/tizen/tizen-controls.ts:12]: Typing won't change import paths. I think the problem is that CommercialControls is not a class so you can't use it as a base for extending.

Why not make CommercialControls a class so that it can extend this class? It does have some functionality related to ads skip button UI that we may need in certain scenarios.

Comment (mamaddox) [reply src/js/view/controls/tizen/tizen-controls.ts:12]: I definitely should have been more clear about typing. What I meant by that was when I had started typing Controls before, I had thought about creating an abstract BaseControls class that could be used by new controls components. So, in this case, TizenControls would import/extend BaseControls as opposed to importing/extending Controls, which I think might solve the issue.

But, from what I could tell, I think the larger issue with this is that CommercialControls instantiates a Controls component.
https://github.com/jwplayer/jwplayer-commercial/blob/master/src/js/view/controls/controls.js#L12
This is what was overriding the TizenControls. Instead of explicitly using the OS Controls here, I tried using the controls component set by the controls-loader (ControlsLoader.controls) and it creates a kind of circular dependency for Tizen. 
TizenControls extends/imports Controls (which is loaded as CommercialControls) and then CommercialControls creates a new instance of TizenControls.
I think converting CommercialControls to a class could possibly result in the same issue since CommercialControls would extend TizenControls and TizenControls would extend the imported CommercialControls.

I'm also curious as to how CommercialControls could be loaded because it seems like importing/using them via OS Controls causes a sort of circular dependency. But, if TizenControls doesn't use OS Controls, CommercialControls wouldn't be loaded from the path resolution. Maybe a commercial controls-loader would be needed and CommercialControls would become like a decorator?

Comment (mamaddox) [reply src/js/view/controls/tizen/tizen-controls.ts]: Converting CommercialControls to a class resolved this! @robwalch 🎉

File: src/js/view/controls/tizen/tizen-controlbar.ts [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+import TimeSlider from 'view/controls/components/timeslider';
+import CustomButton from 'view/controls/components/custom-button';
+import button, { Button } from 'view/controls/components/button';
+import { SimpleTooltip } from 'view/controls/components/simple-tooltip';
+import { toggleClass } from 'utils/dom';
+import type { PlayerAPI } from 'types/generic.type';
+import type ViewModel from 'view/view-model';
+import type NextUpToolTip from 'view/controls/nextuptooltip';
+
+type ControlbarElement = HTMLElement | Button | TimeSlider | any;
```
Comment (zetagame) [inline src/js/view/controls/tizen/tizen-controlbar.ts:12]: This is always any

File: src/js/view/controls/tizen/tizen-controlbar.ts [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+
+function div(classes: string): HTMLElement {
+    const element = document.createElement('div');
+    element.className = `jw-reset ${classes}`;
+    return element;
+}
+
+const appendChildren = (container: HTMLElement, elements: ControlbarElement[]) => {
```
Comment (zetagame) [inline src/js/view/controls/tizen/tizen-controlbar.ts:20]: Are we checking for e.element() because this function takes two types of arguments (JS views and HTML elements)? Rather than allowing it to pass the JS view or the element, we should set this in a way so it's always an element being passed in (single responsibility principle).

File: src/js/view/controls/tizen/tizen-controlbar.ts (hunk 16 above)
Comment (mamaddox) [reply src/js/view/controls/tizen/tizen-controlbar.ts]: Nice catch!

File: src/js/view/controls/tizen/tizen-controlbar.ts (hunk 17 above)
Comment (mamaddox) [reply src/js/view/controls/tizen/tizen-controlbar.ts]: 👍 This gave me an idea to extract the extract out the elements in a new function to pass to appendChildren.

File: src/js/utils/browser.ts (hunk 14 above)
Comment (mamaddox) [reply src/js/utils/browser.ts]: It looks like the only difference between the browser and web application user agent is the inclusion of `Version` and it looks like the web app doesn't have `SamsungBrowser`. I think that might be as specific as we can get.

**Comment by jnatalzia** [comment]
test this please

File: src/css/controls/flags/tizen.less [hunk 18]
```diff
@@ -1,93 +1,239 @@
 @import "../../shared-imports/vars.less";
 
 .jw-tizen-app {
-    // Increasing the size of the svg-icons - icon.less
-    .jw-svg-icon {
-        .square(60px);
+    .jw-title {
+        /* Important is required so the title component is never displayed */
+        /* stylelint-disable-next-line declaration-no-important */
+        display: none !important;
```
[the remark this answers is not in the document]
Comment (mamaddox) [reply src/css/controls/flags/tizen.less]: don't actually think this is needed

File: src/css/controls/imports/settings-menu.less [hunk 19]
```diff
@@ -180,7 +180,7 @@
         }
     }
 
-    .jw-breakpoint-7 &:not(.jw-tizen-app &) {
```
Comment (robwalch) [inline src/css/controls/imports/settings-menu.less:183]: Instead of complicating these selectors (making them more specific which could override other selectors), how about having the tizen controls not apply these breakpoint classes? Just apply the breakpoint(s) class you want in tizen controls by overriding the method that is adding them in controls. This is something that in general we should try to do to simplify our selectors.

File: src/js/view/controls/components/menu/menu.js [hunk 20]
```diff
@@ -65,27 +65,6 @@ export default class Menu extends Events {
             const key = sourceEvent && normalizeKey(sourceEvent.key);
             let focusEl;
 
-
-            switch (sourceEvent.keyCode) {
```
Comment (robwalch) [inline src/js/view/controls/components/menu/menu.js:69]: Why were these moved to tizen? How are these inputs implemented for desktop now?

File: src/css/controls/imports/settings-menu.less (hunk 19 above)
Comment (mamaddox) [reply src/css/controls/imports/settings-menu.less:183]: I felt the same way which is why I am removing all the `not(.jw-tizen-app)` selectors from the css classes. These were initially added for quick styling in the first app. All the tizen styling is in `tizen.less` now and the fullscreen flag is added in tizen-controls.

File: src/js/view/controls/components/menu/menu.js (hunk 20 above)
Comment (mamaddox) [reply src/js/view/controls/components/menu/menu.js:69]: The original implementation was never changed so they still exist for desktop right below. For the first app, this was a quick workaround for navigation through the settings-menu. This is just some cleanup from the first app.

File: src/css/controls/imports/settings-menu.less (hunk 19 above)
Comment (robwalch) [reply src/css/controls/imports/settings-menu.less:183]: Sorry I was commenting on something you fixed, as if it was a new change. ☕☕☕ Nice fix.
